src/colorizer/CanvasOverlay.ts

```diff
@@ -241,34 +254,58 @@ export default class CanvasOverlay implements IRenderCanvas {
   }
 
   public setZoom(zoom: number): void {
-    this.zoomMultiplier = zoom;
+    // TODO: Replace all these checks for specific instances with canvas type (2D/3D)
     if (this.innerCanvas instanceof ColorizeCanvas2D) {
+      this.zoomMultiplier = zoom;
       this.innerCanvas.setZoom(zoom);
     }
     this.render();
   }
 
   public setPan(x: number, y: number): void {
-    this.panOffset.set(x, y);
     if (this.innerCanvas instanceof ColorizeCanvas2D) {
+      this.panOffset.set(x, y);
       this.innerCanvas.setPan(x, y);
     }
     this.render();
   }
 
   public async setParams(params: RenderCanvasStateParams): Promise<void> {
+    const prevParams = this.params;
     this.params = params;
-    // Inner canvas will re-render itself when the params are set.
-    await this.innerCanvas.setParams(params);
+
+    // If the dataset has changed types, construct and initialize the inner
+    // canvas.
+    let hasUpdatedCanvasParams = false;
+    if (hasPropertyChanged(params, prevParams, ["dataset"])) {
+      if (params.dataset) {
+        if (params.dataset.has2dFrames() && this.innerCanvasType !== CanvasType.CANVAS_2D) {
+          this.innerCanvasType = CanvasType.CANVAS_2D;
+          await this.setCanvas(this.innerCanvas2d);
+          hasUpdatedCanvasParams = true;
+        } else if (params.dataset.has3dFrames() && this.innerCanvasType !== CanvasType.CANVAS_3D) {
+          this.innerCanvasType = CanvasType.CANVAS_3D;
+          if (!this.innerCanvas3d) {
+            this.innerCanvas3d = new ColorizeCanvas3D();
+          }
+          await this.setCanvas(this.innerCanvas3d);
+          hasUpdatedCanvasParams = true;
+        }
+      }
+    }
```

Comment:
I wonder if the above code block (lines 282-293, everything inside of `if params.dataset`) could/should be inside its own function somehow.